### ISSUE TITLE
how works createShare?

### ISSUE THREAD

**Issue by Elevencode** [nextcloud/neon#572]
I put file in storage: `await _nextcloudClient.webdav.put(file.bytes!, 'storage/images/${file.name}');`
then I try create share:
```
final imageUrl = await _nextcloudClient.filesSharing.shareapi.createShare(
          path: 'storage/images/${file.name}',
          permissions: 1,
        );
```
but I get an error:

`Image upload failed: FilesSharingApiException(statusCode: 404, headers: {cache-control: no-cache, no-store, must-revalidate, connection: Keep-Alive, content-length: 312, content-security-policy: default-src 'none';base-uri 'none';manifest-src 'self';frame-ancestors 'none', content-type: application/json; charset=utf-8, date: Tue, 15 Aug 2023 10:45:11 GMT, expires: Thu, 19 Nov 1981 08:52:00 GMT, feature-policy: autoplay 'none';camera 'none';fullscreen 'none';geolocation 'none';microphone 'none';payment 'none', keep-alive: timeout=5, max=100, pragma: no-cache, referrer-policy: no-referrer, server: Apache/2.4.41 (Ubuntu), x-content-type-options: nosniff, x-frame-options: SAMEORIGIN, x-permitted-cross-domain-policies: none, x-request-id: vwEfLh2sgzNp4zp7rJqR, x-robots-tag: noindex, nofollow, x-xss-protection: 1; mode=block}, body: {"ocs":{"meta":{"status":"failure","statuscode":404,"message":"Invalid path, file or directory does not exist"},"data":[]}})`

what I do wrong?

**Comment by provokateurin** [comment]
I'm not sure what is wrong, https://github.com/nextcloud/server/blob/89f66a5a4b7e384de3610afba8beea05a16e4f8f/apps/files_sharing/lib/Controller/ShareAPIController.php#L648 is responsible for this and it just checks the path in your user's folder. I don't see something wrong in your code, did you check that the file definitely exists?

Btw if you upload a file you can use `putFile` which is more memory efficient because it doesn't read the whole file into memory at once (though that might not matter if you know your files are always very small).

**Comment by Elevencode** [comment]
@provokateurin yes, file exists. I tried to delay between `.put()` and .`createShare()` but it didn't help.

Thank you for `putFile`, I'll use this :)

**Comment by provokateurin** [comment]
Oh haha I think I see the problem. The parameter is called path but inside the client code we have a variable named path that is used for the path of the URL. I remember seeing this problem in other clients as well. @Leptopoda do you think it is enough to prefix our internal variables with `_`? Technically there could still be a parameter called `_path`.

**Comment by Elevencode** [comment]
@provokateurin Yeah, btw, I saw that too, but I thought it was used for queryParams

**Comment by Leptopoda** [comment]
> Technically there could still be a parameter called _path.

While openapi can do so we wouldn't be able for the generated spec as we use named parameters and they can't start with an underscore.

I also think that prefixing every variable with an underscore is the way to go :+1: 
Another attribute to think of is the class level `rootClient`.

**Comment by provokateurin** [comment]
That's what I thought as well, at least we will get broken code as a warning (and if that ever happens we could change it so that the named parameter doesn't have the underscore).

**Comment by Elevencode** [comment]
So what should I do guys? :) Just wait? How long?

**Comment by provokateurin** [comment]
Yeah, we need to fix it. Please be patient, we are working on this in our free time. You could also try to fix it yourselves in the generator if you feel comfortable enough.

**Comment by Elevencode** [comment]
Why we can't just rename client variable (in constructor)?

**Comment by provokateurin** [comment]
The rootClient field is not the problem. In this case the local `path` variable is because it clashes with the `path` parameter (but this also applies to all other local variables).

**Comment by Elevencode** [comment]
I mean do it in `createShare(collectionPath: ...)`

**Comment by provokateurin** [comment]
The name of the parameter in the API is called path and we try to stay as close as possible to that (except for some names that are not compatible with dart). We can't just add a random prefix (unless it's something like a simple `$`) because people might look at the official documentation (not the package documentation) and won't find the exact parameter in our code.

**Comment by Elevencode** [comment]
Ok, understood!

**Comment by Leptopoda** [comment]
> Ugh prepending an `_` doesn't work as they get removed by `dart fix` :grimacing:. We could also always prefix the named variables with `$` but that is pretty ugly imo. We could also only do that for the variable names that are used internally, but that is risky as it could break if we forget to update the list of forbidden names at some point. @Leptopoda what do you think?

Will dart fix also remove it when there is a naming conflict?

This sounds like a bug in the quick fix action on the linter/analyser side (or whatever handles it).

**Comment by provokateurin** [comment]
> Will dart fix also remove it when there is a naming conflict?

Good idea, I didn't check that.

**Comment by provokateurin** [comment]
Ok that does work, the internal variable gets renamed to `path0`, but i think that is fine as all the code is still valid that way. I don't like generating code that gets optimized away later in 99.9% of the cases, but I guess there is no other way.

**Comment by Leptopoda** [comment]
I might have already mentioned it but that's exactly the type of issue I'd hope #175 to solve/not have in the first place

**Comment by provokateurin** [comment]
I wonder how they fix that :thinking:

**Comment by Leptopoda** [comment]
Probably private var and disable the lint?

I mean we could also do that. I don't think the lint is any useful in our case

**Comment by Elevencode** [comment]
Thanks guys!

**Comment by provokateurin** [comment]
i checked what chopper does and they use $ as the prefix. that is not save as there could be a valid parameter with a $ in the beginning, but maybe they check that such thing doesn't happen :woman_shrugging:
